Pre-flight checks: Add post build action step to trigger TC build

brbrr opened the pull request:
Adds a new step to a Build workflow to trigger TeamCity build for the PR. 

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
pd5faL-eW-p2

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Since the whole integration is in place already, you can see the added status check, which should include the link to the Phab diff and the tests status.
* One can re-run the build action to see the updated status check

brbrr committed:
test

brbrr committed:
test

brbrr committed:
test

brbrr committed:
test

brbrr committed:
test

brbrr committed:
test

brbrr committed:
test

brbrr committed:
test

brbrr committed:
Merge branch 'trunk' into test-tc-integration

brbrr committed:
test

brbrr committed:
test

brbrr committed:
changelog

brbrr committed:
Merge branch 'trunk' into test-tc-integration

brbrr committed:
Merge branch 'trunk' into test-tc-integration

brbrr committed:
Add new build step

brbrr committed:
Add conditional logic

brbrr committed:
Merge branch 'trunk' into test-tc-integration

brbrr committed:
try single quotes

brbrr committed:
Merge branch 'test-tc-integration' of github.com:Automattic/jetpack into test-tc-integration

brbrr committed:
decode JSON string

brbrr committed:
Actually send request to trigger the build

brbrr committed:
revert some test changes

brbrr committed:
remove more unnecessary changes

brbrr committed:
update changelog entry

brbrr committed:
Merge branch 'trunk' into test-tc-integration

brbrr committed:
Merge branch 'trunk' into test-tc-integration

brbrr committed:
Merge branch 'trunk' into test-tc-integration

brbrr commented on .github/workflows/build.yml:
```diff
@@ -130,6 +130,18 @@ jobs:
+          curl -v --fail -L \
+            --url "https://mc.a8c.com/github-post-push-callback.php?jetpack-tc-trigger" \
+            --form-string "pr-number=$PR" \
+            --form-string "sha=$GITHUB_SHA" \
```
Hm, apparently, $GITHUB_SHA is not what we need. For the latest commit, it used `40dfde5e762fd683d619027a4d6c727284910b1e` which isn't a part of this PR:

https://github.com/Automattic/jetpack/commit/40dfde5e762fd683d619027a4d6c727284910b1e

anomiex replied:
For a PR event, I think you'll need to use `github.event.pull_request.head.sha`, like https://github.com/Automattic/jetpack/blob/test-tc-integration/.github/workflows/linting.yml#L284 does.

GITHUB_SHA should work for a push event.

brbrr replied:
Oh, right. I guess we are not interested in push events, so I'll use the suggested approach without determining which event it is.

anomiex replied:
We are interested in push events. That's the type of event triggered when a PR is merged into trunk.

anomiex commented on .github/workflows/build.yml:
```diff
@@ -130,6 +130,18 @@ jobs:
+          PR: ${{ github.event.number }}
+        run: |
+          curl -v --fail -L \
+            --url "https://mc.a8c.com/github-post-push-callback.php?jetpack-tc-trigger" \
```
I wonder if we should make the URL a secret too, so it's not shown here and gets hidden in the logs. Since we have the thing that complains if people just paste such a URL into a comment.

In that case we might also need to drop the `-v` option here too, so curl doesn't print the host part and IP in its logs.

brbrr replied:
Addressed in [c9e5aac](https://github.com/Automattic/jetpack/pull/26616/commits/c9e5aac9269211455c97a38c42e3ea99b3d6143d)

anomiex commented:
It might be useful for the response from the curl to include a bread crumb to find the run in case the check doesn't get reported back. No URL, obviously, but something like "Scheduled run 96 (see pd5faL-eW-p2)" (where pd5faL-eW-p2 explains how to find "run 96") would be enough to point people in the right direction.

> One can re-run the build action to see the updated status check

That would mean waiting for the whole build to run again. If we want to be able to re-run just the check from GitHub (e.g. if TC chokes somehow), we should make the curl be a separate job so only that job needs to be re-run.

brbrr committed:
Multiple small tweaks:
- Replace URL with the secret;
- Remove verbosity flag;
- Use the correct way to find commit SHA;

brbrr committed:
Add a hint on how to find triggered build

brbrr commented:
> One can re-run the build action to see the updated status check

The original quote was part of testing instructions, which is there to test "changing status", not the expected workflow when the actual check/tests are failing.

> If we want to be able to re-run just the check from GitHub

There should not be a need to manually re-run the action/workflow, since the only thing that this trigger does, is triggers a quite simple TeamCity job, that effectively prepares the diff, in which tests would be running. So if the status check on the PR is failing, one would need to look into the tests in a diff, not into the initially triggered TC job.

brbrr committed:
use proper sha param

kraftbj committed:
Merge branch 'trunk' into test-tc-integration

anomiex commented on .github/workflows/build.yml:
```diff
@@ -130,6 +130,22 @@ jobs:
+            --form-string "sha=$SHA" \
+            --form-string "secret=$SECRET")
+
+            echo "The TeamCity build has been triggered. Details: /viewQueued.html?itemId=$response"
```
Complete nitpick, the "echo" should align with line 141. 
```suggestion
          echo "The TeamCity build has been triggered. Details: /viewQueued.html?itemId=$response"
```

anomiex commented on projects/plugins/jetpack/class.jetpack.php:
```diff
@@ -706,6 +706,7 @@ private function __construct() {
 
 		add_filter( 'plugins_url', array( 'Jetpack', 'maybe_min_asset' ), 1, 3 );
 		add_action( 'style_loader_src', array( 'Jetpack', 'set_suffix_on_min' ), 10, 2 );
+
```
We'll want to get rid of the change here before merging, of course. And the changelog entry too.

kraftbj commented on .github/workflows/build.yml:
```diff
@@ -130,6 +130,22 @@ jobs:
+      - name: Trigger TeamCity build
+        if: ${{ fromJSON(steps.changed.outputs.projects)['plugins/jetpack'] == true }}
+        env:
+          SECRET: ${{ secrets.JETPACK__GITHUB_TEAMCITY_TOKEN }}
```
It looks like this might be a personal token @brbrr (watching a build->TC->phab occur now). Not a blocker for merge since the secret can be updated separately.

brbrr replied:
No, this is a secret that is used to verify the request. There are a bunch of other personal tokens that were used, but they'll be updated to shared one by devops.

kraftbj reviewed (dismissed):
Pretty cool. Confirmed the PR triggered TC build. Confirmed that TC took the PR and created a diff for Phab. Watched Phab run all of the WP.com tests and report success back to TC which reported it back here.

It'll be fun to create a PR that would intentionally break wp.com to see the failure condition, but that all seems separate from this code to trigger the run.

anomiex commented on .github/workflows/build.yml:
```diff
@@ -130,6 +130,22 @@ jobs:
+          SECRET: ${{ secrets.JETPACK__GITHUB_TEAMCITY_TOKEN }}
+          URL: ${{ secrets.TEAMCITY_TRIGGER_ENDPOINT }}
+          PR: ${{ github.event.number }}
+          SHA: ${{ github.event.pull_request.head.sha }}
```
One easy way to fix the trunk-push issue might be to do like this.
```suggestion
          PR: ${{ github.event_name == 'push' && github.ref_name || github.event.number }}
          SHA: ${{ github.event_name == 'push' && github.sha || github.event.pull_request.head.sha }}
```
That would result in "pr-number" being "trunk" (or possibly something like "jetpack/branch-11.5") for pushes.

anomiex reviewed (commented):
This change looks ok for pull requests, but when this is merged the added step will fail to supply a PR number or SHA when the trunk build runs.

On the TeamCity side, though, I see that it apparently triggers `jetpack-downloader` immediately and winds up testing the artifact from the previous commit instead of the one just built. One way to fix that might be to poll https://betadownload.jetpack.me/jetpack-branches.json looking for an entry with the corresponding "pr" field and the matching "commit" sha (and then similarly then run for a trunk push). Keep in mind that it may be several minutes thanks to having to wait for the "Test plugin upgrades" job before the betadownload site can actually fetch the artifact.

Or I suppose we could add a "--sha" option to the downloader script, which would cause it to wait for that sha to be available, if that makes more sense to you.

brbrr committed:
Update .github/workflows/build.yml

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

brbrr committed:
Move TC trigger step into a separate job

brbrr committed:
Apply brad suggestion on push events support

brbrr committed:
revert all the testing changes

anomiex commented:
Looks like the remaining issues in this PR have been addressed. I think the only thing remaining is the issue on the TeamCity side where it's running before the betadownload site has downloaded the new artifact.

brbrr committed:
Merge branch 'trunk' into test-tc-integration

brbrr committed:
Merge branch 'test-tc-integration' of github.com:Automattic/jetpack into test-tc-integration

anomiex reviewed (commented):
I think we're good on this end now.

I tried to check the latest build job to see if the issue with it testing the old artifact was fixed, but I got an error "You do not have enough permissions to access build with id: 8859814". :slightly_frowning_face:

Also it'd be nice if we could get DevOps to adjust H145 to exclude the diffs created for these checks before merging this, so we don't get spammed every time the tests run. Maybe by excluding any submitted by tc-phab-bot? Or they might have a better idea. OTOH, I suppose I could just set up a mail filter on my end if we don't want to wait.

brbrr committed:
Merge branch 'trunk' into test-tc-integration

sdixon194 reviewed (approved):
Approved so we can get this moving, we can iterate on getting those emails ignored after.

anomiex commented:
I did not see an email about a Phab diff created from https://github.com/Automattic/jetpack/actions/runs/3329956654/jobs/5508053694, nor do I see anything likely when searching Phab. Unfortunately TC still gives me a "you do not have enough permissions" error when trying to view the build referenced in the log so I can't see what might have gone wrong there.

brbrr commented:
The herald rule was updated, so it won't assign crew for review. The status check does includes a diff is, that might be useful. I'll look into the TC access issues a bit more.
